## Pandas III Metodos

In [1]:
import pandas as pd

1) Cargar el dataset (Pista Si hacemos la carga del dataset sin el parámetro encoding nos puede dar error (puede que no os de error, si es así menos problemas!). Aquí tenéis una lista de los encodings disponibles en pandas. Buscad por los encoding de tipo ISO.)

In [2]:
df = pd.read_csv('files/attacks.csv', index_col= 0)

2) Haced una exploración básica del DataFrame:

- ¿Cuántas filas y columnas tiene el DataFrame?

In [3]:
df.shape

(6114, 21)

Nuestro DataFrame tiene 6114 filas y 21 columnas.

- Mostrad las primeras cinco filas y las últimas cinco filas del DataFrame.

In [4]:
df.head()

Case Number    Year        Type    Country             Area  \
0  2018.06.25  2018.0     Boating        USA       California   
1  2018.06.18  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

   ...         Species           Investigator or Source  \
0  ...      White shark                R. Collier, GSAF   
1  ...              NaN  K.McMurray, TrackingSharks.com   
2  ...              NaN  K.McMurray, TrackingSharks.com   
3  ...        2 m shark                  B. Myatt, GSAF   
4  ...  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order  Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0          NaN         NaN  
1    2018.06.18         6302.0          NaN         NaN  
2    2018.06.09         6301.0          NaN         NaN  
3    2018.06.08         6300.0          NaN         NaN  
4    2018.06.04         6299.0          NaN         NaN  

[5 rows x 21 columns]

In [5]:
df.tail()

Case Number    Year          Type    Country               Area  \
6127  1804.02.26.R  1804.0          Boat  AUSTRALIA    New South Wales   
6128    1803.05.17  1803.0  Sea Disaster        USA     South Carolina   
6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA  Western Australia   
6130  1802.04.13.R  1802.0    Unprovoked      INDIA                NaN   
6131  1801.12.18.R  1801.0      Provoked        NaN                NaN   

                              Location                         Activity  \
6127    Georges Head, off Port Jackson                              NaN   
6128                    Off Charleston                              NaN   
6129  Hamelin Harbour, at Faure Island                              NaN   
6130                               NaN                              NaN   
6131                               NaN  Standing on landed shark's tail   

                                 Name Sex   Age  ...   Species   \
6127                             boat  NaN  NaN  ...        NaN   
6128                    Captain Jones    M  NaN  ...        NaN   
6129  M. Lefevre & a sailor (rescuer)    M  NaN  ...        NaN   
6130                              NaN  NaN  NaN  ...        NaN   
6131                 Stephen Pettigew    N  NaN  ...  12' shark   

                                 Investigator or Source  \
6127                         Sydney Gazette, 2/26/1804    
6128                            Evening Post, 6/13/1803   
6129  F.  Peron ref in G.P. Whitley (Fishes of Austr...   
6130                  Evening Post (New York) 4/13/1802   
6131                       The Evening Post, 12/18/1801   

                                  pdf  \
6127  1804.02.26-boat-GeorgesHead.pdf   
6128             1803.05.17-Jones.pdf   
6129           1803.03.00-Lefevre.pdf   
6130           1802.04.13.R-India.pdf   
6131       1801.12.18.R-Pettigrew.pdf   

                                           href formula  \
6127  http://sharkattackfile.net/spreadsheets/pdf_di...   
6128  http://sharkattackfile.net/spreadsheets/pdf_di...   
6129  http://sharkattackfile.net/spreadsheets/pdf_di...   
6130  http://sharkattackfile.net/spreadsheets/pdf_di...   
6131  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
6127  http://sharkattackfile.net/spreadsheets/pdf_di...  1804.02.26.R   
6128  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.05.17   
6129  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.03.00   
6130  http://sharkattackfile.net/spreadsheets/pdf_di...  1802.04.13.R   
6131  http://sharkattackfile.net/spreadsheets/pdf_di...  1801.12.18.R   

     Case Number.2 original order  Unnamed: 22 Unnamed: 23  
6127  1804.02.26.R          176.0          NaN         NaN  
6128    1803.05.17          175.0          NaN         NaN  
6129    1803.03.00          174.0          NaN         NaN  
6130  1802.04.13.R          173.0          NaN         NaN  
6131  1801.12.18.R          172.0          NaN         NaN  

[5 rows x 21 columns]

- Mostrad 2 filas al azar del DataFrame.

In [6]:
df.sample(2)

Case Number    Year        Type    Country        Area  \
1097  2009.08.25  2009.0  Unprovoked        USA  California   
4893  1935.12.23  1935.0    Provoked  AUSTRALIA    Victoria   

                                        Location  Activity  \
1097  Terramar Beach, Carlsbad, San Diego County  Swimming   
4893            150 yards from the pier at Lorne  Cruising   

                                          Name Sex   Age  ...  \
1097                            Bethany Edmund    F   22  ...   
4893  16' motor launch owned by A. & E. Norton  NaN  NaN  ...   

                             Species         Investigator or Source  \
1097  White shark, 5' to 6' juvenile               R. Collier, GSAF   
4893                4.3 m [14'] shark  V.M. Coppleson (1958), p.183   

                            pdf  \
1097      2009.08.25-Edmund.pdf   
4893  1935.12.23-NortonBoat.pdf   

                                           href formula  \
1097  http://sharkattackfile.net/spreadsheets/pdf_di...   
4893  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1097  http://sharkattackfile.net/spreadsheets/pdf_di...    2009.08.25   
4893  http://sharkattackfile.net/spreadsheets/pdf_di...    1935.12.23   

     Case Number.2 original order  Unnamed: 22 Unnamed: 23  
1097    2009.08.25         5206.0          NaN         NaN  
4893    1935.12.23         1410.0          NaN         NaN  

[2 rows x 21 columns]

- ¿Cuál es el nombre de las columnas del DataFrame?

In [7]:
columnas= df.columns
columnas

Index(['Case Number', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

- ¿Cuáles son los principales estádisticos del DataFrame? Mostrad por separado las variables categóricas y las variables numéricas.

In [8]:
df.describe() # Variables numericas

Year  original order
count  6114.000000     6114.000000
mean   1968.831207     3237.445044
std      43.745920     1769.652633
min    1801.000000      172.000000
25%    1946.000000     1706.250000
50%    1980.000000     3236.500000
75%    2005.000000     4768.750000
max    2018.000000     6303.000000

In [9]:
df.describe(include= 'object') # Variables categoricas

Case Number        Type Country     Area  \
count         6113        6110    6074     5706   
unique        6113           8     201      791   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        4438    2200     1025   

                                Location Activity  Name  Sex    Age  \
count                               5626     5603  5914  5564  3448   
unique                              4020     1484  5118     6   156   
top     New Smyrna Beach, Volusia County  Surfing  male     M    17   
freq                                 162      966   511  4939   154   

             Time     Species  Investigator or Source                     pdf  \
count        2934         3409                   6098                    6114   
unique        366         1529                   4837                    6105   
top     Afternoon  White shark         C. Moore, GSAF  1898.00.00.R-Syria.pdf   
freq          183          161                     95                       2   

                                             href formula  \
count                                                6113   
unique                                               6104   
top     http://sharkattackfile.net/spreadsheets/pdf_di...   
freq                                                    2   

                                                     href Case Number.1  \
count                                                6114          6114   
unique                                               6099          6113   
top     http://sharkattackfile.net/spreadsheets/pdf_di...    1952.08.04   
freq                                                    4             2   

       Case Number.2   Unnamed: 22 Unnamed: 23  
count           6114             1           2  
unique          6114             1           2  
top       2018.06.25  stopped here      Teramo  
freq               1             1           1

- ¿Cuál es el porcentaje de valores nulos en cada columna?

In [10]:
for i in columnas:
    resultado = (df[i].isnull().sum()/len(df[i]))*100
    print(f"{i} {round(resultado,2)} %")

Case Number 0.02 %
Year 0.0 %
Type 0.07 %
Country 0.65 %
Area 6.67 %
Location 7.98 %
Activity 8.36 %
Name 3.27 %
Sex  9.0 %
Age 43.6 %
Time 52.01 %
Species  44.24 %
Investigator or Source 0.26 %
pdf 0.0 %
href formula 0.02 %
href 0.0 %
Case Number.1 0.0 %
Case Number.2 0.0 %
original order 0.0 %
Unnamed: 22 99.98 %
Unnamed: 23 99.97 %


- Sacad los valores únicos y su frecuencia de las variables numéricas. Almacenad los resultados en un diccionario. Uno para cada variable.

In [12]:
dic_unicos ={} #Creamos un diccionario con variables numericas
for i in df:
    
    if df.dtypes[i] == 'object':
        pass
    else:
         dic_unicos.update({i:(df[i].value_counts())})

In [13]:
dic_unicos.keys() #Verificamos que columnas trajo

dict_keys(['Year', 'original order'])

In [16]:
# convertimos el diccionario en tres uno por cada variable numerica como pedia la consigna.


dic_year = dict(dic_unicos['Year'])
dic_original = dict(dic_unicos['original order'])



- Cambiad el nombre de las columnas para que todas estén en minúsculas y reemplazad los espacios por _.

In [17]:
cambio_columnas = {col :  col.lower().strip().replace(' ','_') for col in df.columns}

In [20]:
df1 = df.rename(columns = cambio_columnas) #inplace = True

In [21]:
df1.columns  # verificamos los cambios

Index(['case_number', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'time', 'species',
       'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

- Eliminad las columnas de Unnamed: 22 y Unnamed: 23, Case Number.1, Case Number.2, original order, Investigator or Source, pdf, href formula.

In [22]:
df1.drop(['investigator_or_source', 'pdf', 'href_formula','case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'], axis=1, inplace=True)

In [23]:
df1.columns # Verificamos los cambios

Index(['case_number', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'time', 'species', 'href'],
      dtype='object')

- Establece la columna Case Number como índice del DataFrame.

In [24]:
df1.set_index(['case_number'], inplace= True)
df1.head(1)

year     type country        area                     location  \
case_number                                                                     
2018.06.25   2018.0  Boating     USA  California  Oceanside, San Diego County   

             activity         name sex age   time      species  \
case_number                                                      
2018.06.25   Paddling  Julie Wolfe   F  57  18h00  White shark   

                                                          href  
case_number                                                     
2018.06.25   http://sharkattackfile.net/spreadsheets/pdf_di...

- Guarda el csv que no tiene esas columnas para seguir usándolo mañana.

In [25]:
df1.to_csv('files/attacks1.csv')